File: labs/m5-agent-sdk/typescript/solution/src/tools.ts
import { createSdkMcpServer, tool } from "@anthropic-ai/claude-agent-sdk";
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { z } from "zod";

const TICKETS = "tickets.json";
export const TICKET_TOOL = "mcp__tracker__create_ticket";

export function appendTicket(title: string, severity: string, file: string, line: number): string {
  const tickets: unknown[] = existsSync(TICKETS) ? JSON.parse(readFileSync(TICKETS, "utf8")) : [];
  const id = `AST-${String(tickets.length + 1).padStart(4, "0")}`;
  tickets.push({ id, title, severity, file, line, created: new Date().toISOString() });
  writeFileSync(TICKETS, JSON.stringify(tickets, null, 2));
  return id;
}

const createTicket = tool(
  "create_ticket",
  "File a bug ticket in the Astronomy Shop tracker for one confirmed finding. Returns the new ticket id.",
  { title: z.string(), severity: z.string().describe("HIGH | MEDIUM | LOW"), file: z.string(), line: z.number().int() },
  async (args) => {
    const severity = args.severity.toUpperCase();
    if (!["HIGH", "MEDIUM", "LOW"].includes(severity)) {
      return { content: [{ type: "text", text: `severity must be HIGH, MEDIUM or LOW, got ${args.severity}` }], isError: true };
    }
    const id = appendTicket(args.title, severity, args.file, args.line);
    return { content: [{ type: "text", text: `created ${id} for ${args.file}:${args.line}` }] };
  },
);

export const trackerServer = createSdkMcpServer({ name: "tracker", version: "1.0.0", tools: [createTicket] });

File: labs/m5-agent-sdk/python/solution/bughunter/tools.py
"""Step 3 — custom tool served from an in-process SDK MCP server. M6 reuses append_ticket() via labs/shared/tickets.py."""
import json
import time
from pathlib import Path
from typing import Any

from claude_agent_sdk import create_sdk_mcp_server, tool

TICKETS = Path("tickets.json")          # lives in the directory you run bughunter from


def append_ticket(title: str, severity: str, file: str, line: int) -> str:
    tickets = json.loads(TICKETS.read_text()) if TICKETS.exists() else []
    ticket_id = f"AST-{len(tickets) + 1:04d}"
    tickets.append({"id": ticket_id, "title": title, "severity": severity, "file": file, "line": line,
                    "created": time.strftime("%Y-%m-%dT%H:%M:%S")})
    TICKETS.write_text(json.dumps(tickets, indent=2))
    return ticket_id


@tool("create_ticket",
      "File a bug ticket in the Astronomy Shop tracker for one confirmed finding. Returns the new ticket id.",
      {"title": str, "severity": str, "file": str, "line": int})
async def create_ticket(args: dict[str, Any]) -> dict[str, Any]:
    severity = str(args.get("severity", "")).upper()
    if severity not in {"HIGH", "MEDIUM", "LOW"}:
        # is_error tells Claude the call failed (and why) without killing the loop
        return {"content": [{"type": "text", "text": f"severity must be HIGH, MEDIUM or LOW, got {args.get('severity')!r}"}],
                "is_error": True}
    ticket_id = append_ticket(args["title"], severity, args["file"], int(args["line"]))
    return {"content": [{"type": "text", "text": f"created {ticket_id} for {args['file']}:{args['line']}"}]}


tracker_server = create_sdk_mcp_server(name="tracker", version="1.0.0", tools=[create_ticket])
TICKET_TOOL = "mcp__tracker__create_ticket"   # mcp__<key in mcp_servers>__<tool name>
